Handle nonconformant group block size

hschimke opened the pull request:
This PR contains two changes to support slightly non-conformant files found in the wild.

1. In some cases the length of the GroupStart block is set such that the block only contains the name of the group. When that is the case the `Group::parse()` function cannot add all member colors to the group. Colors which appear after the GroupStart block but before a GroupEnd block should be included in the group, so the change attempts to capture those colors after the initial GroupStart block is parsed and insert them as they are found. This still requires a GroupEnd block to be found, and so new error logic had to be inserted to handle this situation.
2. The length of a GroupEnd block is always zero. However, in files found in the wild the length is often specified in the four bytes following the group type. This causes those files to be parsed incorrectly, even though the actual logic should be the same (a zero length block is still a zero length block whether the size is explicitly stated or not). The PR adds logic to detect those zero-length explicitly sized blocks and skip the following four bytes when it is safe to do so.

This change addresses the second half of PF #1 .

I had initially hoped that I could solve this issue by changing the logic for `Group::parse()` but since the size is not known ahead of time it would have meant passing the entire buffer into the function and duplicating the logic found in `read_ase()`.  I suspect there is a more elegant solution to the "skip unnecessary GroupEnd block size bytes" problem, but the one presented here works with all the tests and with the ASE files that I am trying to parse with it.

hschimke committed:
feat: include enhanced error messages

BREAKING CHANGE: applications which interact with the ASEError type directly will need to be modified to handle the additional error values.

hschimke committed:
test: add unit tests for new error types

hschimke committed:
docs: update documentation for new error types

hschimke committed:
feat: handle files with nonconformant group block sizes

In some files found in the wild GroupStart blocks have a size which only contains the name of the group, this causes colors in that group to not be parsed. This change fixes that special case.

In addition, the change fixes an issue where some files feature a specified block size of 0 for GroupEnd blocks.

hschimke committed:
fix: formatting and exports for new error types

hschimke committed:
docs: add documentation for group block logic

hschimke committed:
fix: fix skipping logic for specified group-end blocks

hschimke committed:
fix: check for empty blocks with is_empty()

FineFindus committed:
fix: remove double backticks

FineFindus committed:
build: increase major version

hschimke committed:
chore: Merge branch 'master' into handle_nonconformant_group_block_size

hschimke commented:
This cleanly merges now. It's a pretty major change to how the `read_ase` function works, though.

hschimke committed:
test: add test for groups with explicitly set GroupEnd length

FineFindus commented on src/lib.rs:
```diff
@@ -76,12 +77,45 @@ pub fn read_ase<T: std::io::Read>(mut ase: T) -> Result<(Vec<Group>, Vec<ColorBl
+            safe_to_skip = true;
+            skipped = 0;
+
+            0
```
```suggestion
            safe_to_skip = true;
            skipped = 0;
            0
```

FineFindus commented on src/lib.rs:
```diff
@@ -266,6 +331,37 @@ mod tests {
+        );
+        let block = ColorBlock::new("name".to_owned(), ColorValue::Gray(0.5), ColorType::Normal);
+        let input_ase_bytes = create_ase(vec![group.clone()], vec![block.clone()]);
+        let mut modified_ase_bytes = vec![0; 0];
+        modified_ase_bytes.extend_from_slice(&input_ase_bytes[..128]);
+        modified_ase_bytes.extend_from_slice(&[0; 4]);
+        modified_ase_bytes.extend_from_slice(&input_ase_bytes[128..]);
```
Please add a comment/illustration explaining what exactly is done here, since it can be hard to read this at first glance.

FineFindus commented on src/lib.rs:
```diff
@@ -266,6 +331,37 @@ mod tests {
+        assert!(res.is_ok());
+        let res = res.unwrap();
+        assert_eq!(res, (vec![group], vec![block]));
+        assert_eq!(res.0.first().unwrap().name, "group name".to_owned());
+        assert_eq!(res.1.first().unwrap().name, "name".to_owned());
```
Why check the names again? They should already be compared in the assert above.

FineFindus commented on src/lib.rs:
```diff
@@ -90,22 +124,53 @@ pub fn read_ase<T: std::io::Read>(mut ase: T) -> Result<(Vec<Group>, Vec<ColorBl
+                // if the parsed block has any sub-blocks then it has already been built
+                // and only a group-end block may follow it. Otherwise we are free to
+                // add colors as they appear until a group-end block is encountered.
+                group_hold = if block.blocks.is_empty() {
+                    types::GroupHold::HoldingBuilding
+                } else {
+                    blocks_to_read += 1;
+                    types::GroupHold::HoldingBuilt
+                };
```
```suggestion
                group_hold = if block.blocks.is_empty() {
                    GroupHold::HoldingBuilding
                } else {
                    blocks_to_read += 1;
                    GroupHold::HoldingBuilt
                };
```

`types::GroupHold` is already imported. Please also adjust the other usages.

FineFindus reviewed (changes requested):
> I had initially hoped that I could solve this issue by changing the logic for Group::parse() but since the size is not known ahead of time it would have meant passing the entire buffer into the function and duplicating the logic found in read_ase(). I suspect there is a more elegant solution to the "skip unnecessary GroupEnd block size bytes" problem, but the one presented here works with all the tests and with the ASE files that I am trying to parse with it.

Since the only difference is the length, have you considered parsing in the `Group::parse()` function until an end block is found? Passing a reference to the buffer is fine, if it enables better code.

hschimke committed:
test: include tests for sized groups and sized groups and specified group end

chore: this also includes changes requested for comments, formatting, and type scope.

hschimke commented:
I'm going to build another version that attempts to move all group color parsing into the Group::parse() function. I'll present it as a separate PR.

The two tests I added should help with that, as they capture the specific issues originally presented.

hschimke commented:
Hmm, I've been working on the Group::parse method a little bit this morning and I'm going to keep poking at it, but I have a suspicion that it makes more sense to leave Group::parse() as is.

I have two reasons for this:
1. It logically makes more sense. In the case where the GroupStart block is sized such that it holds the entire set of colors for the group then you are logically parsing a single "block" with Group::parse(), while if the GroupStart block is sized only for the name of the group then really the remaining blocks are independent, though they do happen to be part of that group.
2. The initial bit of code that I've written seems messier to me, with more duplicated logic. I'm worried that the more places we do the same thing the more likely it is to break.

That said, I don't want to discount it. I'll play with it a bit more and see if I can make it make more sense.

FineFindus commented:
Thanks!

> That said, I don't want to discount it. I'll play with it a bit more and see if I can make it make more sense.

Please remember to take some time off, not just work on this :)
Ping me when you're done and haven't found a better way, I'll wait with the next release until then.

hschimke commented:
@FineFindus I finished a version that handles the majority of the parsing logic in the `Group::parse()` function. I honestly don't know if it's cleaner that way. I'll leave it up to you to decide which makes more sense in the context of the library.

I didn't make a PR for it, but you can check out the code on the branch below, if you prefer it I'll make a PR.

https://github.com/hschimke/adobe-swatch-exchange-rs/tree/enhanced_group_parse

> Please remember to take some time off, not just work on this :)

No worries, this is my hobby/entertainment! I've been playing with this to relax in the mornings while I drink my tea.

FineFindus commented:
> No worries, this is my hobby/entertainment! I've been playing with this to relax in the mornings while I drink my tea.

It's also mine, but I wanted to make sure you didn't feel pressured to finish it :)

> I finished a version that handles the majority of the parsing logic in the Group::parse() function. I honestly don't know if it's cleaner that way. I'll leave it up to you to decide which makes more sense in the context of the library.

Yes, after reading the code, I tend to agree. Spreading the logic over two different functions is probably not a good idea. Maybe I'll take another stab at it when I have a little more time.

Thanks for the contributions, it was a pleasure working with you.

hschimke commented:
> Thanks for the contributions, it was a pleasure working with you.

Likewise! If you ever get the urge to spend too much time thinking about [barcodes](https://github.com/rxing-core/) let me know!
